### ISSUE TITLE
Cannot delete read-only files on Windows

### ISSUE THREAD

**Issue by andreynering** [gogits/gogs#1738]
1. Create a repository with a name (let's say "myrepository")
2. Push something to "myrepository"
3. Delete "myrepository" on Gogs
4. Try to create another repository called "myrepository"

It returns 500. I think this happens because the repository folder is not deleted, just the database registry. I tried to rename the folder and could create the repository again.

**Comment by Unknwon** [comment]
Thanks your feedback!

1. What is your Gogs version?
2. Which database are you using?

**Comment by andreynering** [comment]
Sorry, I forgot these information.

Gogs version: 0.6.15
Windows Server 2012
PostgreSQL 9.4

2015-10-06 15:35 GMT-03:00 无闻 <notifications@github.com>:

> Thanks your feedback!
>
>    1. What is your Gogs version?
>    2. Which database are you using?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gogits/gogs/issues/1738#issuecomment-145958136>.
>



-- 
Andrey Nering
andrey.nering@gmail.com
Skype: andrey.nering

**Comment by Unknwon** [comment]
Thanks for following up!

> It returns 500. I think this happens because the repository folder is not deleted, just the database registry. 

Can you 100% reproduce this problem by doing the same thing again?

And check if there is any error in the log(as far as I know, Gogs only prints an error if there is something wrong with delete files, could be a permission issue).

**Comment by andreynering** [comment]
You are right: the log show a permission issue on deleting the repository.

Log here: https://gist.github.com/andreynering/99e1db9fc9a6c9a2f5b8

Also, is pull requests supposed to be working already? It is giving me
error 500 on merge, too. But I didn't check the log.

2015-10-06 15:48 GMT-03:00 无闻 <notifications@github.com>:

**Comment by Unknwon** [comment]
> You are right: the log show a permission issue on deleting the repository.
Log here: https://gist.github.com/andreynering/99e1db9fc9a6c9a2f5b8

Great, but I can't help with permission issue so may be reloaded somewhere that Gogs can have full control of.


> Also, is pull requests supposed to be working already? It is giving me
error 500 on merge, too. But I didn't check the log.

It's kind of working, with very limited functionality, I'm still working on it, so things of PR are subject to change.

**Comment by andreynering** [comment]
Thank you for the attention.

I will let you know if I find more information.

**Comment by Unknwon** [comment]
No problem :smile:

**Comment by Unknwon** [comment]
Any updates?

**Comment by andreynering** [comment]
Sorry, but I still didn't find a solution for this.

**Comment by Unknwon** [comment]
@andreynering 

I think you already pointed out the problem:

> You are right: the log show a permission issue on deleting the repository.

That is something Gogs can't fix itself.

**Comment by andreynering** [comment]
I know what is the problem. I don't know how to fix it. ¯\_(ツ)_/¯

**Comment by Unknwon** [comment]
I don't know either.... 

And I was asking:

> Can you 100% reproduce this problem by doing the same thing again?

**Comment by andreynering** [comment]
> Can you 100% reproduce this problem by doing the same thing again?

Yes, just tried again today on latest release.

**Comment by Unknwon** [comment]
@andreynering I still insist that you need to check something like delete permission, maybe you put repos in a admin-level directories and so Gogs can't delete anything unless it is ran as admin.

**Comment by andreynering** [comment]
Tomorrow I will try change the folder permission. I already tried another
Windows user and had the same issue. I'll post here if it work or not.

**Comment by Unknwon** [comment]
@andreynering OK, maybe try to put in ~/Downloads directory other than home directory (Gogs default).

**Comment by andreynering** [comment]
I was running Gogs with `COMPUTERNAME$` as user.

Tried give everybody "full control" permission on the repositories folder.
No luck.

Then tried run gogs as `MYUSER` and copied the folder to
`MYUSER/Downloads`. No luck.

Both printed the same error on log: permission denied.

**Comment by Unknwon** [comment]
Can you rename a repository?

**Comment by andreynering** [comment]
Yes.

**Comment by Unknwon** [comment]
Please post the complete error message here.

**Comment by Unknwon** [comment]
Also, can you help check the repo files are not `read-only`?

**Comment by Unknwon** [comment]
Found something relative... https://github.com/golang/go/issues/9606

**Comment by andreynering** [comment]
> Please post the complete error message here for permission denied with latest version you tested.

```
2015/11/03 08:16:28 [W] delete repository files[D:\xxx\gogs-repositories\andreynering\teste.git]: remove D:\xxx\gogs-repositories\andreynering\teste.git\objects\19\8d5d252f4ac328120d476a5592b94aa13677c8: Access is denied.
```

> Also, can you help check the repo files are not `read-only`?

Yes, all files are read-only.

> Found something relative...

Hmm... Maybe Gogs should write files as `0666`?

**Comment by Unknwon** [comment]
> Yes, all files are read-only.

Combine with golang/go#9606 , we know t he problem now... 

But objects files are wrote by git itself, not by Gogs...

**Comment by haliphax** [comment]
I am suffering from this, as well, and it might prevent us from using gogs as our enterprise source control solution. :( I can't tell my supervisor we need a *nix machine for the service -- we only run Windows.

**Comment by Unknwon** [comment]
@haliphax too bad... any special hack/command could possibly force delete files on Windows (when run Gogs as admin)?

**Comment by Unknwon** [comment]
Sounds like a working workaround?

> cmd := exec.Command("cmd.exe", "/C", "rmdir", "/S", "/Q", dirPathWithWindowsSeparators)

**Comment by haliphax** [comment]
I am using it as admin, and yes, using the `/s /q` flag with `rmdir` would do the trick.

**Comment by Unknwon** [comment]
@haliphax I'll try it. Thanks!

**Comment by Tim91084** [comment]
This ever make it in?  Having the same issue myself - I have to manually delete the repo from the working directory.  Just wanted to note another (perhaps more prevalent) usecase.

1.) Fork repo "ABC"
2.) Do work, submit PR ...
3.) Delete Forked repo
4.) Realize you need to make another change
5.) Try to fork repo again -> 500 error

**Comment by Unknwon** [comment]
@Tim91084 does [this comment](https://github.com/gogits/gogs/issues/1738#issuecomment-164065321) work on your system?

**Comment by Tim91084** [comment]
@Unknwon sorry for my ignorance, but what should I do with that line?  Do you want me to recompile the gogs source with that change, or just test the command in command prompt?

If recompiling the source, could you point me in the general direction where I would make that change?

**Comment by Unknwon** [comment]
@Tim91084 No.. i just want your help to verify that command works on your Windows shell/xyz.

**Comment by Tim91084** [comment]
```
D:\>rmdir test
The directory is not empty.

D:\>rmdir /?
Removes (deletes) a directory.

RMDIR [/S] [/Q] [drive:]path
RD [/S] [/Q] [drive:]path

    /S      Removes all directories and files in the specified directory
            in addition to the directory itself.  Used to remove a directory
            tree.

    /Q      Quiet mode, do not ask if ok to remove a directory tree with /S

D:\>rmdir /S /Q test
```

The /S /Q args seem to mitigate the issue or removing a directory that is not empty.  If that is the root cause of this issue, it's likely that adding those arguments will fix it.

**Comment by Unknwon** [comment]
Can anyone follow this thread help test `develop` branch as we merged a fix 2570?

**Comment by Unknwon** [comment]
@andreynering Any time you could try the fix? https://github.com/golang/go/commit/2ffb3e5d905b5622204d199128dec06cefd57790

**Comment by andreynering** [comment]
@Unknwon 

I compiled Gogs with the master branch of Go compiler and without [the workaround I did to fix this](https://github.com/andreynering/gogs/blob/cc2964cf0b36c4041c6b55e88f7a84af462643fb/models/admin.go#L81). I can confirm it is fixed, it can delete read-only files/folders. I recreated the same repository + wiki multiples times without any error (I tried with Go 1.6 right before and it still errored without the workaround).

But we still have to wait to remove the workaround. The fix will be in Go 1.7, and Gogs supports, I think, the 3 lasts Go versions?
